src/VecSim/algorithms/svs/svs_tiered.h

```diff
@@ -0,0 +1,732 @@
+          updateJobWaitTime(
+              tiered_index_params.specificParams.tieredSVSParams.updateJobWaitTime == 0
+                  ? 1000 // default wait time: 1ms
+                  : tiered_index_params.specificParams.tieredSVSParams.updateJobWaitTime),
+          uncompletedJobs(this->allocator) {
+        this->journal.reserve(this->updateJobThreshold * 2);
+    }
+
+    int addVector(const void *blob, labelType label) override {
+        int ret = 0;
+        auto svs_index = GetSVSIndex();
+        if (this->getWriteMode() == VecSim_WriteInPlace) {
+            // Use the frontend parameters to manually prepare the blob for its transfer to the SVS
+            // index.
+            auto storage_blob = this->frontendIndex->preprocessForStorage(blob);
+            std::scoped_lock lock(this->updateJobMutex, this->mainIndexGuard);
+            return svs_index->addVectors(storage_blob.get(), &label, 1);
+        }
+        bool index_update_needed = false;
+        {
+            std::scoped_lock lock(this->flatIndexGuard, this->mainIndexGuard, this->journal_mutex);
+            ret = this->frontendIndex->addVector(blob, label);
+            ret = std::max(ret - svs_index->deleteVectors(&label, 1), 0);
+            journal.emplace_back(label, true);
+            index_update_needed = this->backendIndex->indexSize() > 0 ||
+                                  this->journal.size() >= this->updateJobThreshold;
+        }
+
+        if (index_update_needed) {
+            scheduleSVSIndexUpdate();
+        }
+
+        return ret;
+    }
+
+    int deleteVector(labelType label) override {
+        int ret = 0;
+        auto svs_index = GetSVSIndex();
+        if (this->getWriteMode() == VecSim_WriteInPlace) {
+            assert([&] {
+                std::shared_lock<std::shared_mutex> flat_lock(this->flatIndexGuard);
+                return !this->frontendIndex->isLabelExists(label);
+            }());
+
+            std::scoped_lock lock(this->updateJobMutex, this->mainIndexGuard);
+            return svs_index->deleteVectors(&label, 1);
+        }
+
+        bool label_exists = [&] {
+            std::shared_lock<std::shared_mutex> flat_lock(this->flatIndexGuard);
+            return this->frontendIndex->isLabelExists(label);
+        }();
+
+        bool index_update_needed = false;
+        if (label_exists) {
+            std::scoped_lock lock(this->flatIndexGuard, this->mainIndexGuard, this->journal_mutex);
+            if (this->frontendIndex->isLabelExists(label)) {
+                ret = this->frontendIndex->deleteVector(label);
+                assert(ret == 1 && "unexpected deleteVector result");
+            }
+            ret += svs_index->deleteVectors(&label, 1);
+            assert(ret < 2 && "deleteVector: vector duplication in both indices");
+            journal.emplace_back(label, false);
+            index_update_needed = this->journal.size() >= this->updateJobThreshold;
```

Comment:
can you please explain when index_update_needed and what does the update include?

Reply:
Removed.